MTV-1211 Max concurrent virtual machine migrations

RichardHoch opened the pull request:
MTV 2.7

Resolves https://issues.redhat.com/browse/MTV-1211 by adding information about `controller_max_vm_inflight` (shown in the UI as *Max concurrent virtual machine migrations*).

Previews:

- https://file.corp.redhat.com/rhoch/max_concur_vm/html-single/#configuring-mtv-operator_mtv  [first bullet in list and first item in Table 3.1]
- https://file.corp.redhat.com/rhoch/max_concur_vm/html-single/#max-concurrent-vms_mtv [new section]
- https://file.corp.redhat.com/rhoch/max_concur_vm/html-single/#mtv-settings_mtv [first item in Table 4.1]

nunzy1 commented:
Your call as to what you incorporate. :-)   

I might change:

"Maximum number of virtual machines (VMs) or disks per plan that can be migrated simultaneously."
To:
"Maximum number of virtual machines (VMs) or disks per plan that can migrate simultaneously."   

(---unless “can be” means that you may decide not to and have that option)


"The maximum number of disks that can be transferred simultaneously." 
To:  
"The maximum number of disks that can transfer simultaneously." 

"The maximum number of VMs that can be migrated simultaneously." 
To: 
"The maximum number of VMs that can migrate simultaneously."


"…can be transferred simultaneously. In these migrations, the disks are migrated in parallel. This means that if the combined number of disks that you want to migrate is greater than the value of the setting, additional disks wait until the queue is free without regard for whether a VM has completely finished being migrated."  
To:
 "…can be transferred simultaneously. In these migrations, the disks migrate in parallel. If the combined number of disks that you want to migrate is greater than the value of the setting, additional disks migrate only once the queue is free, without regard for whether a VM has finished migrating."

"Once any of them has migrated, the 16th disk can be migrated, whether or not all the disks on VM A and VM B have finished migrating." 
To:
 "Once any of them has migrated, the 16th disk migrates, whether or not all the disks on VM A and VM B have finished migrating."

  (--  unless “can be migrated” means that it’s an option to migrate the disk)

anarnold97 commented on documentation/modules/max-concurrent-vms.adoc:
```diff
@@ -0,0 +1,11 @@
+[id="max-concurrent-vms_{context}"]
+= A detailed look at controller_max_vm_inflight
+
+For {rhv-full}, {osp}, vmWare warm migrations _or_ remote cold migrations, the setting `controller_max_vm_inflight` (shown in the UI as *Max concurrent virtual machine migrations*) specifies the maximum number of _disks_ that can be transferred simultaneously. In these migrations, the disks are migrated in parallel. This means that if the combined number of disks that you want to migrate is great ...
```
```suggestion
For {rhv-full}, {osp}, VMware warm migrations _or_ remote cold migrations, the setting `controller_max_vm_inflight` (shown in the UI as *Max concurrent virtual machine migrations*) specifies the maximum number of _disks_ that can be transferred simultaneously. In these migrations, the disks are migrated in parallel. This means that if the combined number of disks that you want to migrate is greater than the value of the setting, additional disks wait until the queue is free without regard for whether a VM has completely finished being migrated.
```

Remember the correct format for [VMware](https://www.vmware.com/products/cloud-infrastructure/vsphere-foundation)

![image](https://github.com/user-attachments/assets/ed61a986-ea2c-4d79-9d87-7e39cc2cca72)

anarnold97 commented on documentation/modules/mtv-settings.adoc:
```diff
@@ -14,7 +14,11 @@ If you have Administrator privileges, you can access the *Overview* page and cha
-|The maximum number of VMs per plan that can be migrated simultaneously
+a|*Varies with provider as follows*:
+
+* {rhv-full}, {osp}, vmWare warm migrations _or_ remote cold migrations: The maximum number of disks that can be transferred simultaneously.
```
```suggestion
* {rhv-full}, {osp}, VMware warm migrations _or_ remote cold migrations: The maximum number of disks that can be transferred simultaneously.
```

RichardHoch replied:
Fixed.

anarnold97 commented on documentation/modules/mtv-settings.adoc:
```diff
@@ -14,7 +14,11 @@ If you have Administrator privileges, you can access the *Overview* page and cha
+a|*Varies with provider as follows*:
+
+* {rhv-full}, {osp}, vmWare warm migrations _or_ remote cold migrations: The maximum number of disks that can be transferred simultaneously.
+* OVA or vmWare local cold migrations: The maximum number of VMs that can be migrated simultaneously.
```
```suggestion
* OVA or VMware local cold migrations: The maximum number of VMs that can be migrated simultaneously.
```

anarnold97 commented on documentation/modules/configuring-mtv-operator.adoc:
```diff
@@ -37,7 +37,11 @@ spec:
-|The maximum number of VMs per plan that can be migrated simultaneously.
+|a|*Varies with provider as follows*:
+
+* {rhv-full}, {osp}, vmWare warm migrations _or_ remote cold migrations: The maximum number of disks that can be transferred simultaneously.
```
```suggestion
* {rhv-full}, {osp}, VMware warm migrations _or_ remote cold migrations: The maximum number of disks that can be transferred simultaneously.
```

RichardHoch commented:
@sgratch Please review this PR. Thanks.

sgratch commented:
@RichardHoch @mnecas 
Isn't it worth renaming the parameter to reflect its meaning to something like:
_Max concurrent virtual machine **or disks** migrations?_

RichardHoch commented:
@sgratch  Changing the name of the parameter is up to @mnecas as far as I am concerned. But if we do it, I would suggest  `controller_max_vm_disk_inflight` and Max concurrent virtual machine or disk migrations  [drop the "s" in "disks"  -- english is weird].  Does the PR look OK otherwise?

sgratch commented on documentation/modules/configuring-mtv-operator.adoc:
```diff
@@ -37,7 +37,13 @@ spec:
-|The maximum number of VMs per plan that can be migrated simultaneously.
+a|*Varies with provider as follows*:
+
+* {rhv-full}, {osp}, VMware warm migrations _or_ remote cold migrations: The maximum number of disks that {project-short} can transfer simultaneously.
```
For **all** occurrences of  **remote cold migrations**:
I suggest to explain what does "remote migration" mean  and therefore rephrase to something like:
 **cold migrations to a remote OpenShift cluster**

sgratch commented on documentation/modules/configuring-mtv-operator.adoc:
```diff
@@ -37,7 +37,13 @@ spec:
+
+* {rhv-full}, {osp}, VMware warm migrations _or_ remote cold migrations: The maximum number of disks that {project-short} can transfer simultaneously.
+
+* OVA or VMware local cold migrations: The maximum number of VMs that {project-short} can migrate simultaneously.
```
For **all** occurrences of   **local cold migrations**:
I suggest to explain what does "local migration" mean and therefore rephrase to something like:
 **cold migrations to a local OpenShift cluster**

mnecas replied:
It does not apply for **all** local cold migrations but afaik only for virt-v2v related migrations.

sgratch commented on documentation/modules/max-concurrent-vms.adoc:
```diff
@@ -0,0 +1,16 @@
+
+Example: If the value of the setting is 15, and VM A has 5 disks, VM B has 5 disks, and VM C has 6 disks, all the disks except for the 16th begin migrating at the same time. Once any of them has migrated, the 16th disk can be migrated, whether or not all the disks on VM A and VM B have finished migrating.
+
+However, for OVA or for local VMware local cold migrations, the setting specifies the maximum number of _VMs_ that {project-short} can migrate simultaneously, meaning that all additional disks must wait until at least one VM has been completely migrated.
```
```suggestion
However, for OVA or for VMware local cold migrations, the setting specifies the maximum number of _VMs_ that {project-short} can migrate simultaneously, meaning that all additional disks must wait until at least one VM has been completely migrated.
```

sgratch reviewed (commented):
LGTM except the comments below.
But let's wait for @mnecas final approval as well.

sgratch commented:
>  Changing the name of the parameter is up to @mnecas as far as I am concerned. But if we do it, I would suggest `controller_max_vm_disk_inflight` and Max concurrent virtual machine or disk migrations [drop the "s" in "disks" -- English is weird].

Sounds good to me to change it. @mnecas ?

mnecas commented:
> > Changing the name of the parameter is up to @mnecas as far as I am concerned. But if we do it, I would suggest `controller_max_vm_disk_inflight` and Max concurrent virtual machine or disk migrations [drop the "s" in "disks" -- English is weird].
> 
> Sounds good to me to change it. @mnecas ?

I am a bit afraid that we would lose the previous settings during the upgrade.
I don't want to change it to the "or" as we might even change the logic to always transfer disks.

RichardHoch commented:
@sgratch @mnecas I made the change Sharon suggested. Please review this PR again. Thanks.

sgratch commented:
@RichardHoch  did you miss those 2 comments ? :)
https://github.com/kubev2v/forklift-documentation/pull/580#discussion_r1836951105
https://github.com/kubev2v/forklift-documentation/pull/580#discussion_r1836953157,

RichardHoch commented:
@sgratch @mnecas I made some more changes to the PR. Please review them. Thanks

duduvaa commented:
The 'controller_max_vm_inflight' parameter is related to simultaneous migration.
While using VMware as the source provider:
COLD migration - VMs  per ESXi host that can migrate simultaneously
WARM migration - Disks  per ESXi host that can migrate simultaneously

anarnold97 commented on documentation/modules/configuring-mtv-operator.adoc:
```diff
@@ -37,7 +37,13 @@ spec:
-|The maximum number of VMs per plan that can be migrated simultaneously.
+a|*Varies with provider as follows*:
+
+* {rhv-full} migrations, {osp} migrations, VMware warm migrations, _or_ cold VMware migrations to a remote {ocp} cluster: The maximum number of disks that {project-short} can transfer simultaneously.
+
+* OVA migrations or cold VMware migrations using `virt-v2v` to a local {ocp} cluster: The maximum number of VMs that {project-short} can migrate simultaneously.
```
```suggestion
* For {rhv-full} migrations, {osp} migrations to a remote {ocp} cluster: The maximum number of disks that {project-short} can transfer simultaneously.
* For OVA migrations using `virt-v2v` to a local {ocp} cluster: The maximum number of VMs that {project-short} can migrate simultaneously.
*  For VMware simultaneous migration:
** Cold migration: VMs for each ESXi host that can migrate simultaneously
** Warm migration: Disks for each ESXi host that can migrate simultaneously 
```

@rhoch @duduvaa - please can you have a look at this and see if it would be ok? i am trying to synthesize what has been approved with David's comments

RichardHoch replied:
Thanks, @anarnold97 
@duduvaa Martin said the for VM cold migrations, there was a difference between local and remote. Do you see a difference in your tests?

mnecas replied:
For cold remote we use the CNV CDI so the max in flight references the number of disks.

dgur replied:
@duduvaa can you check regarding Cold remote?

sgratch commented:
> @sgratch I think I included both your changes -- let me know if I missed something.

LGTM

duduvaa commented:
@anarnold97 

LGTM

RichardHoch commented:
@duduvaa You replied LGTM to Andy's suggestion, but what about cold remote? Martin commented "For cold remote we use the CNV CDI so the max in flight references the number of disks" Does that fit your test results? 
That is, what do you say to:  

*  For VMware simultaneous migration:

** Cold migration:

*** To local {virt}: VMs for each ESXi host that can migrate simultaneously
*** To remote {virt}: Disks for each ESXi host that can migrate simultaneously  

** Warm migration: Disks for each ESXi host that can migrate simultaneously

duduvaa commented:
@RichardHoch ,
I'm trying to set my environment for remote cold migration.
I'll update

duduvaa commented:
@RichardHoch ,

I tested cold migration to remote env.
As Martin said, "max_vm_flight" references the number of disks per ESXi host.

Richard, as you wrote
** Cold migration:

*** To local {virt}: VMs for each ESXi host that can migrate simultaneously
*** To remote {virt}: Disks for each ESXi host that can migrate simultaneously

RichardHoch commented:
@mnecas @duduvaa Please review. What about OpenShift Virtualization as a source? VMs or disks?

dgur commented:
@RichardHoch  I believe  you asked about Redhat Virtualization/ Openstack not OCP-V

RichardHoch commented:
> @RichardHoch I believe you asked about Redhat Virtualization/ Openstack not OCP-V

@duduvaa No, I meant OpenShift Virtualization, aka CNV, because it's a source provider, but we didn't define whether the parameter applies to disks or VMs when CNV is the source provider.

duduvaa commented:
> > @RichardHoch I believe you asked about Redhat Virtualization/ Openstack not OCP-V
> 
> @duduvaa Actually, I did mean OpenShift Virtualization, aka CNV, because it's a source provider, but we didn't define whether the parameter applies to disks or VMs when CNV is the source provider.

@fabiand , @mnecas 
Could you relate to Richard's comment?

dgur commented:
In MTV, cnv cannot be source provider only destination

On Mon, 6 Jan 2025, 13:10 David Vaanunu, ***@***.***> wrote:

> @RichardHoch <https://github.com/RichardHoch> I believe you asked about
> Redhat Virtualization/ Openstack not OCP-V
>
> @duduvaa <https://github.com/duduvaa> Actually, I did mean OpenShift
> Virtualization, aka CNV, because it's a source provider, but we didn't
> define whether the parameter applies to disks or VMs when CNV is the source
> provider.
>
> @fabiand <https://github.com/fabiand> , @mnecas
> <https://github.com/mnecas>
> Could you relate to Richard's comment?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubev2v/forklift-documentation/pull/580#issuecomment-2572888700>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHGJQIVDD2GAMNMH26C76LD2JJQCBAVCNFSM6AAAAABQ6X725GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNZSHA4DQNZQGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

fabiand commented:
OCP can be both - src + dst.

@mnecas mentioned that max inflight for CNV is about inflight VMs.

mnecas commented:
We are using the `VirtualMachineExport` to transfer the VMs between clusters/providers.
https://kubevirt.io/user-guide/storage/export_api/

RichardHoch commented:
@mnecas I have added the information for CNV as provider. Please review the PR again. Thanks.

mnecas commented on documentation/modules/max-concurrent-vms.adoc:
```diff
@@ -0,0 +1,16 @@
+[id="max-concurrent-vms_{context}"]
+= Configuring the controller_max_vm_inflight label
+
+For {rhv-full} migrations, {osp} migrations, VMware warm migrations, or cold VMware migrations to a remote {ocp} cluster the setting `controller_max_vm_inflight` (shown in the UI as *Max concurrent virtual machine migrations*) specifies the maximum number of _disks_ that {project-first} can transfer simultaneously. In these migrations, {project-short} migrates the disks in parallel. This means th ...
```
I did more reading and found out that the rhv and osp also do it per VM.
I think the code is not correct and should be changed but this is the current status so I wanted to correct myself.
So the per disk applies only to remote vmware or warm vmware migrations.

RichardHoch commented:
@mnecas I made the changes you suggested. Please review this PR again.

anarnold97 commented on documentation/modules/max-concurrent-vms.adoc:
```diff
@@ -0,0 +1,27 @@
+
+* For all migrations except OVA or VMware migrations, the label specifies the maximum number of _disks_ that {project-first} can transfer simultaneously. In these migrations, {project-short} migrates the disks in parallel. This means that if the combined number of disks that you want to migrate is greater than the value of the setting, additional disks must wait until the queue is free, without r ...
++
+Example: If the value of the label is 15, and VM A has 5 disks, VM B has 5 disks, and VM C has 6 disks, all the disks except for the 16th begin migrating at the same time. Once any of them has migrated, the 16th disk can be migrated, whether or not all the disks on VM A and VM B have finished migrating.
```
```suggestion
Example: If the value of the label is 15, and VM A has 5 disks, VM B has 5 disks, and VM C has 6 disks, all the disks except for the 16th disk start migrating at the same time. Once any of them has migrated, the 16th disk can be migrated, even though not all the disks on VM A and the disks on VM B have finished migrating.
```

pedantic comments:

* personally i would include `disk` as it makes `begin` appear to be a noun, including `disk` might make it a wee bit clearer (not minimalism but i think it reads a bit better)
* personally i would use `even though not... ` instead of `whether or not` 
* personally i would spell out it is the `VM A disks and the VM B disks` ... same as before

anarnold97 commented on documentation/modules/max-concurrent-vms.adoc:
```diff
@@ -0,0 +1,27 @@
+
+* For OVA migrations, the label specifies the maximum number of _VMs_ that {project-short} can migrate simultaneously, meaning that all additional disks must wait until at least one VM has been completely migrated.
++
+Example: If the value of the label is 2, and VM A has 5 disks, VM B has 5 disks, and VM C has 6 disks, all the disks on VM C must wait until either all the disks on VM A or on VM B have finished migrating.
```
```suggestion
For example, if the value of the label is 2, and VM A has 5 disks, VM B has 5 disks, and VM C has 6 disks. All the disks on VM C must wait to migrate until either all the disks on VM A or on VM B to finish migrating.
```

RichardHoch commented:
@anarnold97: Please give this PR another review.

nunzy1 commented:
https://file.corp.redhat.com/rhoch/max_concur_vm/html-single/#configuring-mtv-operator_mtv [first bullet in list and first item in Table 3.1]

In the preview, in the first item in table 3.1, middle column, the lead-in is bold.  Is that what you intended?  Also, the left-hand column - do you want those to be all bold?  The left-hand column of 4.1 is not bold (I know they are “label” versus “Setting”, so perhaps our standards call for this.)  

Just above 3.1, “Spec: label: value” looks large.  If this follows our standards, cool.  Just wanted to point it out.


https://file.corp.redhat.com/rhoch/max_concur_vm/html-single/#max-concurrent-vms_mtv [new section]

This bullet point includes VM in italics; the others do not.  “For OVA migrations, the label specifies the maximum number of VMs that MTV…”  

https://file.corp.redhat.com/rhoch/max_concur_vm/html-single/#mtv-settings_mtv [first item in Table 4.1]

That first item in Table 4.1 also includes a bold lead-in for the middle column, but not for subsequent entries into the middle column.  Also, at the end of the first column, “+ See [Configuring the](https://file.corp.redhat.com/rhoch/max_concur_vm/html-single/#max-concurrent-vms_mtv)…”    The +See caught my eye.

RichardHoch commented:
@nunzy1 1. FYI -- In the previews, items with backticks appear to be bolded, but they aren't in production. You're right about "Varies with provider," which i fixed. 
2. The italicization of VMs in the section is intended; "disks" is italicized in the bullet above. I think that "disks" vs. "VMs" should be italicized. Do you think that's overkill or that the emphasis is lost in the text? Either reason would be a good reason to drop the italics -- what do you think?
3.  Fixed both. Thanks.
